magento/magento2
# Use per-store or per-website skin if available in the error application

### Description (*)
When Magento is unable or unwilling to process a request (eg, because maintenance mode has been enabled) the request is forwarded to the application within `pub/errors/`. This uses a different set of theme files to the main website (intentionally), but is not flexible when dealing with multi-store or multi-website configurations.

This is not necessarily a problem when all the stores/website share a common theme and language. (For example, British & American & Australian stores with the same brand but different currencies / localisation will all be fine using a shared skin/theme in English for `pub/errors/`.) However, when there are multiple distinct brands being served from a single Magento instance or when different languages are being used, a single skin for `pub/errors/` is not suitable.

We have implemented a similar approach to what is being proposed here on websites for our merchants with good results. Here we are using the `MAGE_RUN_TYPE` and `MAGE_RUN_CODE` environment variables to determine which skin / theme to use in error pages.

Calling `_setSkin()` multiple times is safe. If there is no suitable skin matching the name of the store/website, then these lines have no impact, falling back to other skin-setting mechanisms like `$_GET['skin']` or `pub/errors/local.xml` or `pub/errors/design.xml`.

### Related Pull Requests

- https://github.com/magento/devdocs/pull/9363

### Fixed Issues
*None.* Please feel free to create an issue if you feel this is appropriate.

### Manual testing scenarios (*)
1. Set up Magento with multi-store or multi-website, ensure these have distinct URLs, eg https://de.magento.test/ https://es.magento.test/ https://fr.magento.test/. - see guide here for details: https://devdocs.magento.com/guides/v2.4/config-guide/multi-site/ms_over.html
1. Enable maintenance mode with `bin/magento maintenance:enable`
1. Navigate to each store front. Notice that all serve the same skin for `pub/errors/`.
1. Create a new skin for error pages by copying `pub/errors/local.xml.sample` to `pub/errors/local.xml`, and `pub/errors/default/` to `pub/errors/custom/`, and setting `<skin>custom</skin>` in `pub/errors/local.xml`.
1. Navigate to each store front. Notice that all serve the same skin for `pub/errors/`.
1. Notice no ability to customise the errors per-store nor per-website before the changes here.
1. Apply the changes in this pull request.
1. Create new skins for each store/website as desired. For example, copy `pub/errors/custom/` to `pub/errors/es/` and translate/theme this appropriately.
1. Navigate to each store front. Notice that a different skin is applied to each where this exists, and that where no matching theme/skin exists, the default is used.

### Questions or comments
*None*

### Contribution checklist (*)
